### ISSUE TITLE
Optimize data loading performance

### ISSUE THREAD

**Issue by mathpluscode** [DeepRegNet/DeepReg#302]
# Issue description

For nifti files, now we are loading them in py and provide a generator to TensorFlow.
This generator is sequential, and therefore the loading is sequential?

Maybe we should provide a dataset of file names, then load the data in parallel in TensorFlow.

Guideline: https://www.tensorflow.org/guide/data_performance

It is unclear if this can be useful for H5 data.

## Type of Issue

- [ ] New feature request

**Comment by YipengHu** [comment]
It can be quite complicated and application-dependent! with different buffers shifting the optimal loading/processing/transferring time - i would say, if the tfrecord does not improve a lot - as you tried, we should leave this level of optimisation to Google engineers :)

**Comment by YipengHu** [comment]
> It can be quite complicated and application-dependent! with different buffers shifting the optimal loading/processing/transferring time - i would say, if the tfrecord does not improve a lot - as you tried, we should leave this level of optimisation to Google engineers :)

or users...

**Comment by mathpluscode** [comment]
> It can be quite complicated and application-dependent! with different buffers shifting the optimal loading/processing/transferring time - i would say, if the tfrecord does not improve a lot - as you tried, we should leave this level of optimisation to Google engineers :)

Oh yes, maybe need to try tfrecord #99, it was indeed faster but not that much because of the conv3d layers which took the most of the time in calculation. So agree, maybe it's better to add back the tfrecord support and benchmark the performance ;)

**Comment by mathpluscode** [comment]
Btw, we should activate the profiling to understand what's going on in our runs.
And for the profiling, with a large chance we need the docker file as mentioned in #420.

**Comment by sbb-gh** [comment]
@YipengHu @mathpluscode can we clafify the aim of this issue?  Are we interested in optimizing the dataloading for nifty, or for giving users different options for dataloading?

**Comment by YipengHu** [comment]
> @YipengHu @mathpluscode can we clafify the aim of this issue?  Are we interested in optimizing the dataloading for nifty, or for giving users different options for dataloading?

For the first instance, a brief investigation and estimates of what can be done, what options we have and the likely benefits for each. We will open separate issue(s), if we decide to implement any of these.

**Comment by sbb-gh** [comment]
Do we need to assume the data is in a format?

**Comment by YipengHu** [comment]
> Do we need to assume the data is in a specific format?

Good question. It is still not clear to me if h5 or tfrecord is faster than say nifti in these small minibatch problems... Part of the question for you I guess 🤪

**Comment by mathpluscode** [comment]
> @YipengHu @mathpluscode can we clarify the aim of this issue? Are we interested in optimizing the data loading for nifty, or for giving users different options for data loading?

There are two issues I'd like to address:

1. The original purpose of this issue. Now we are building the dataset from a generator, which sequentially reads multiple files. But in case of multiple large files, we can instead build multiple generators and use `interleave` to accelerate the data loading. For this part, we will need to profile the training process, and with the docker file, it's feasible now.

2. Another way to accelerate is to perform deterministic pre-processing on the dataset and store it into tfrecords, and this ideally should accelerate the training. Earlier this year, I tried that, and it provided 1.5x acceleration I think. This might be linked to #446.

Does it answer your questions? @sbb-gh If you'd like to contribute to any of these two points, feel free to go on and we could arrange a meeting to sync between us.

**Comment by mathpluscode** [comment]
> > Do we need to assume the data is in a specific format?
> 
> Good question. It is still not clear to me if h5 or tfrecord is faster than say nifti in these small minibatch problems... Part of the question for you I guess 🤪

I'd say either the original format (h5 or nifti) or tfrecord. I doubt the value to transform between Nifti and h5?

**Comment by sbb-gh** [comment]
Let's schedule a meeting

**Comment by YipengHu** [comment]
i close this for now and follow up with #459
